tensorflow-core/tensorflow-core-api/pom.xml

```diff
@@ -31,6 +31,11 @@
       <version>${project.version}</version>
       <optional>true</optional> <!-- for compilation only -->
     </dependency>
+    <dependency>
+      <groupId>org.tensorflow</groupId>
+      <artifactId>nio-utils</artifactId>
+      <version>${project.version}</version>
+    </dependency>
```

Comment:
That feels strange. Why do we need this dependency here? Is it just because that we can't have a separate module until we move away from the code generator in C++?

Reply:
`Tensor` are now based on `NdArray` for mapping their memory and allowing direct access to it, so it must depend on `nio-utils`. I'm not sure to understand what is strange, you meant that memory mapping should occur outside the core API?

Reply:
I think it's just the names that are confusing me. When I read something like `tensorflow-core`, I think like this means the actual "core" without dependencies on other modules. Maybe it's just that `tensorflow-nio` should really be named something else and not "TensorFlow"? That might also help with adoption from other projects. That's probably something we should continue on the mailing list you created about that and one of the first things to clear up with guys from MXNet, DL4J, etc.

Reply:
Exactly, that's why I renamed the artifact `tensorflow-nio-utils` to `nio-utils` (though I'm not a huge fan of that name neither so if anyone comes up with a better one...). So the `tensorflow` you've noticed there is just to mention our organization in the group and does not name the artifact itself.

Reply:
`ndarray-core`?  I think `nio-utils` is likely to make people think of `java.nio.*` and as the language moves away from those classes (and they are 15+ years old at this point), calling them "new" is something of a misnomer.

Reply:
Or `ndarray-utils`? Or the other way around `utils-ndarray`? Sounds more Java-ish and in line with the name of the parent artifact

Reply:
yeah so I'm convinced as well, `nio` is out. `ndarray-utils` I like, I understand the logic but it sounds more natural than `utils-ndarray`. Or just `ndarray`? `nd-utils`? 

There are `DataBuffer` as well but they are mainly there to support Nd-arrays.

Reply:
yeah, I don't dislike simply `org.tensorflow:ndarray`, wdyt?

Reply:
I think I prefer `ndarray` then `ndarray-utils`. I think both `ndarray-utils` and `utils-ndarray` seem a little odd given this module is actually the base classes for ndarrays.

Reply:
Ok, just for the sake of picking something, `ndarray` it is, thanks guys!

Reply:
I don't think it belongs in "tensorflow" anyway, whether is goes under `tensorflow-utils` or not is a moot point IMO.

Reply:
We could rename that parent package to `tensorflow-misc` and just put whatever doesn't fit in there. That would kind of make more sense to me...

Reply:
> I don't think it belongs in "tensorflow" anyway

That's a valid point. Actually, we can host that code elsewhere since it does not depend on TF but it is the other way around, like JavaCPP. Should we?

Reply:
Yes, I think everything should be in JavaCPP :) But seriously, this is something that needs to be discussed with people from other projects too.

As for the matter at hand, the name of the artifact, looking back at how Spring does it at https://github.com/spring-projects/spring-framework maybe we shouldn't try to categorize all the modules. `tensorflow-core` has the Bazel build with settings for JavaCPP and everything so that can stay like that, but I think we could have `tensorflow-ndarray` along with `tensorflow-jkeras`, etc, all as main modules under `tensorflow-java`, without a kind of "dump" like `tensorflow-frameworks` or `tensorflow-utils`. That sounds nice to me. In any case, it's probably going to be temporary if we are able at some point to migrate to a version that comes from another repo and that is accepted by other organizations such as Apache MXNet and Eclipse Deeplearning4j as well.

Reply:
I agree, we will group artifacts if there is a need but right now, I cannot think of any other utility library than this one.

Reply:
I think there should be a functional need to group things together. If it's just to put them in abstract categories, it will make it hard to decide which category each module belongs to, possibly without any actual benefits. Maven doesn't even put the parent name in artifact coordinates, so unless we do it explicitly like with `tensorflow-core` -> `tensorflow-core-api`, etc then the benefits are even more marginal.

Reply:
Ok, here is what I ended up to do: I preserved the `tensorflow-utils` name but now it contains the code of the library itself, including `DataBuffer` and `NdArray` APIs. 

So if in the future it happens that we have more small utility classes like these that we would like to share with the world and that are independent from TF runtime, they will end up in this library as well (kind of a Guava for ML). 

There is a lot of presumptions and questions that makes it hard to pick the right name for this library right away, so let's simply rename it in the future if we need to.

Reply:
Sounds good, but let's name it `tensorflow-util` just to be consistent with the package name? Or inversely, let's name the package `org.tensorflow.utils`? I don't have preference for either, I just think that consistency is a good thing to have whenever it makes sense.

Reply:
I think it is fine to have them mismatch here, `tensorflow-utils` sounds more natural for the name of library (as there is more than one utility class in it) while package names are often singular by convention (e.g. `java.util.*`). 

In addition, we don't have a `org.tensorflow.core` package neither in our core artifacts.